src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1135,30 +1137,27 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, Compilation
                         argsToParamsOpt: argsToParams, resultKind: LookupResultKind.Viable, binderOpt: this, type: returnType, hasErrors: gotError);
         }
 
-        private bool IsBindingModuleLevelAttribute()
+        /// <summary>
+        /// Returns false if an implicit 'this' copy will occur due to an instance member invocation in a readonly member.
+        /// </summary>
+        internal bool CheckImplicitThisCopyInReadOnlyMember(BoundExpression receiver, MethodSymbol method, DiagnosticBag diagnostics)
         {
-            if ((this.Flags & BinderFlags.InContextualAttributeBinder) == 0)
+            // For now we are warning only in implicit copy scenarios that are only possible with readonly members.
+            // Eventually we will warn on implicit value copies in more scenarios. See https://github.com/dotnet/roslyn/issues/33968.
+            if (receiver is BoundThisReference &&
+                receiver.Type.IsValueType &&
+                ContainingMemberOrLambda is MethodSymbol containingMethod &&
+                containingMethod.IsEffectivelyReadOnly &&
+                // Ignore calls to base members.
+                TypeSymbol.Equals(containingMethod.ContainingType, method.ContainingType, TypeCompareKind.ConsiderEverything) &&
```

Comment:
Consider doing an assert that the containing method is `System.Object`. For now, the only thing that goes through this path is `struct`s which can only inherit from that, but if `readonly` members were to come to classes they'd likely break this assumption.

Reply:
Maybe that `containingMethod.ContainingType` is either equal to `method.ContainingType` or equal to `System.ValueType`?